### DISCUSSION TITLE
Comparing benches in order to test multiple approaches

### DISCUSSION THREAD

**Discussion by johnbillion** [phpbench/phpbench#1031] [Q&A]
Thanks for your work on PHPBench. It's really been a great help to me on several projects.

I'm working on performance improvements to an application and I'd like to compare two different approaches. This is different to comparing results for regression testing as I'm working on various approaches simultaneously and I'd like to see how their performance compares.

Is this possible to do with PHPBench? At the moment I need to run a benchmark with the first approach, then adjust the `bench*()` method to use the alternative approach, and run it again with the `--ref` flag. I can automate the two test runs but I can't very easily automate changing the code path in the bench method.

Am I thinking about this the wrong way? Any advice would be appreciated.

**Comment by johnbillion** [comment]
Okay after reading through #831 I can see how to do this by introducing a report column which shows a comparison with another subject. I've not yet figured out how to do this when sets are involved (with `@ParamProviders`).

**Comment by dantleech** [comment]
Does the `--report=benchmark_compare` give you anything like what you want?

![image](https://github.com/phpbench/phpbench/assets/530801/f0c63c47-529c-49bf-ace8-dae95f01bdfb)


You need only have a method per thing you want to test

**Reply by johnbillion** [reply]
To be honest not really. I should have mentioned in the original post that I'm using `@ParamProviders` on each of the bench methods. Essentially:

```php
/**
 * @ParamProviders("provideData")
 */
public function benchApproach1( $data ) {
	approach1( $data );
}

/**
 * @ParamProviders("provideData")
 */
public function benchApproach2( $data ) {
	approach2( $data );
}

public function provideData() {
	yield 'first data set' => [ 1, 2, 3 ];
	yield 'second data set' => [ 4, 5, 6 ];
}
```

The `aggregate` and `benchmark_compare` reports show the timings for each subject+set combination, but I'd like to compare each of the sets for the two benches and show the diff:

```
+----------------------------------+---------+-------------+
| subject                          | mode    | improvement |
+----------------------------------+---------+-------------+
| benchApproach1 (first data set)  | 10ms    | 0.00%       |
| benchApproach1 (second data set) | 8ms     | 0.00%       |
| benchApproach2 (first data set)  | 5ms     | -50.00%     |
| benchApproach2 (second data set) | 2ms     | -75.00%     |
+----------------------------------+---------+-------------+
```
